R/api.R

```diff
@@ -442,6 +450,61 @@ nv_reduce_any <- nvl_reduce_any
 #' @export
 nv_reduce_all <- nvl_reduce_all
 
+#' @title Random Numbers
+#' @name nv_rng_bit_generator
+#' @description
+#' generate random bits of desired shape and dtype
+#' @param initial_state state seed
+#' @param rng_algorithm one of 'DEFAULT', 'THREE_FRY', 'PHILOX'
+#' @param dtype datatype of output
+#' @param shape_out output shape
+#' @export
+nv_rng_bit_generator <- nvl_rng_bit_generator
+
+#' @title Random Uniform Numbers
+#' @name nv_runif
+#' @description
+#' generate random uniform numbers
+#' @param initial_state state seed
+#' @param shape_out output shape
+#' @param lower lower bound
+#' @param upper upper bound
+#' @export
+nv_runif <- function(initial_state, shape_out, lower = 0, upper = 1) {
+  rbits <- nv_rng_bit_generator(initial_state = initial_state, "THREE_FRY", "ui64", shape_out = shape_out)
+  lhs <- nv_convert(rbits[[2]], "f64")
+  rhs <- nv_tensor(rep(2^64 - 1, prod(shape_out)), dtype = "f64", shape = shape_out)
```

Comment:
just use a scalar, it will be broadcast. Otherwise huge constant in stablehlo program, makes binary big and compilation slow.

Reply:
Can 2^64 - 1 be exactly represented in R?

Reply:
Use `nv_minval`, `nv_maxval`: https://github.com/r-xla/anvil/blob/d2c99be3d9e3f92a1cf113f1be7dffae530e95fb/R/utils.R#L105C1-L116C39